lib/src/utils/rest_args_with_separator.dart

```diff
@@ -0,0 +1,44 @@
+import 'package:args/args.dart';
+
+/// Returns the "rest" args from [argResults], but with the arg separator "--"
+/// restored to its original position if it was included.
+///
+/// This is necessary because [ArgResults.rest] will _not_ include the separator
+/// unless it stopped parsing before it reached the separator.
+///
+/// The use case for this is a [CompoundTool] that uses the [takeAllArgs] arg
+/// mapper, because the goal there is to forward on the original args minus the
+/// consumed options and flags. If the separator has also been removed, you may
+/// hit an error when trying to parse those args.
+///
+///     var parser = ArgParser()..addFlag('verbose', abbr: 'v');
+///     var results = parser.parse(['a', '-v', 'b', '--', '--unknown', 'c']);
+///     print(results.rest);
+///     // ['a', 'b', '--unknown', 'c']
+///     print(restArgsWithSeparator(results));
+///     // ['a', 'b', '--', '--unknown', '-c']
+List<String> restArgsWithSeparator(ArgResults argResults) {
+  // If no separator was used, return the rest args as is.
+  if (!argResults.arguments.contains('--')) {
+    return argResults.rest;
+  }
+
+  final args = argResults.arguments;
+  final rest = argResults.rest;
+  var rCursor = 0;
+  for (var aCursor = 0; aCursor < args.length; aCursor++) {
+    // Iterate through the original args until we hit the first separator.
```

Comment:
I'm having trouble following this. It seems like the only result of this is setting rCursor. But isn't rCursor just `args.indexOf('--')`?

Reply:
That would be the position of the separator within the original arguments list, which isn't necessarily the same as the position at which the separator should be restored in the `rest` arguments list, since the former excludes flags and options that were parsed by the ArgParser. So for example:

```dart
final argParser = ArgParser()
  ..addOption('output')
  ..addFlag('verbose');
final argResults = argParser.parse(['foo', '--output', 'out', '--verbose', '--', 'bar', '--baz']);
print(argResults.arguments);
// ['foo', '--output', 'out', '--verbose', '--', 'bar', '--baz']
// In this list, separator index is 4
print(argResults.rest);
// ['foo', 'bar', '--baz']
// In this list, we want to restore the separator at index 1
```

Reply:
OK, I get it. Ugh. It'd be better if ArgParser just didn't think it knew about '--' and included it.